Performance optimizations.

phmarek opened the pull request:
Together these changes reduce run time by about 50%, and similarly reduce consing to about 40%.
Please take a look and check whether they work for you; my minimal testing was successful.

The biggest remaining offender is EVALUATE-FEATURE-2 - but I'm not sure I understand
what this is checking, so I'm not up to modifying that.

Thank you!

phmarek committed:
Use an array to count the run-length.

This increases performance as NTH is no longer used.

phmarek committed:
Only deref MATRIX a single time.

phmarek committed:
Declare type MATRIX

phmarek committed:
Avoid using NTH, that amounts to ~6% of total runtime

phmarek committed:
Simplify SYMBOL->PNG.

phmarek committed:
A few types, and TRUNCATE is down to place 7.

phmarek committed:
More declarations.

phmarek commented:
Forgot a bugfix commit, sorry

jnjcc commented:
Thanks for your help!

I have not been active on github lately, will look into it

phmarek committed:
Export the function encode-png-bytes-stream to produce inline PNGs.

phmarek committed:
Fixed function arity by switching to an exported function.

phmarek commented:
ping...

phmarek committed:
Fix for newer SBCL versions; the initform needs to be compatible.

phmarek commented:
Hi!

I just stumbled upon all that again...

Looking at the tests (to find out about the failing test mentioned in #11) I found that your 0de2d8a3877b499a9a0bbb0a9e1247056ae4311e doesn't compile cleanly, the test functions use a wrong function signature ;/


Well, I fixed my `perf` branch for current SBCL (the `:initform` type test change); in 6fd84ed3a770ec2882ef2319225c2bc89ebf4713 you can find a test script with some (old) results.

Should we just disable the failing test if no-one knows what it's about?

phmarek committed:
Optimization (~29% of runtime) in MODULE-PLACEMENT.

A loop generating the same QR PNG 500 times was at:

    Evaluation took:
      3.842 seconds of real time
      3.844585 seconds of total run time (3.772390 user, 0.072195 system)
      ...
      7,659,849,586 processor cycles
      2,697,809,008 bytes consed

With this change we get

    Evaluation took:
      2.739 seconds of real time
      2.745916 seconds of total run time (2.713921 user, 0.031995 system)
      ...
      5,460,421,894 processor cycles
      819,525,456 bytes consed

phmarek commented:
ping?

phmarek committed:
Define MATRIX as SIMPLE-ARRAY, and use ARRAY-DIMENSIONS.

With the benchmark `(time (dotimes (i 2000) (qrcode:encode-symbol "aaa")))` the
timing goes from

    Evaluation took:
      1.543 seconds of real time
      1.545931 seconds of total run time (1.542021 user, 0.003910 system)
      3,075,414,185 processor cycles
      352,981,216 bytes consed

to

    Evaluation took:
      1.167 seconds of real time
      1.167331 seconds of total run time (1.167331 user, 0.000000 system)
      2,327,108,767 processor cycles
      350,175,872 bytes consed

ie. 25% processor cycles less.

jnjcc commented:
WOW, long time no see!

I have been busy at work and have been away from github for sooo long,
will definitely look into this!

Thanks again for your help

phmarek commented:
I'm also pondering to change the bit-lists into a bit-array - or to use bignum support to avoid all the bit-list fiddling at all.

See also https://gist.github.com/jdz/3831dc3b5af5cb76b18b14384216913c.
